File: leetcode.com/linked-list-cycle.py
#!/bin/python3
# https://leetcode.com/problems/linked-list-cycle/

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        slow = head
        fast = head

        while True:
            if fast == None or fast.next == None:
                return False

            fast = fast.next.next
            slow = slow.next

            if fast == slow:
                return True

File: leetcode.com/linked-list-cycle.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

def hasCycle(head)
    slow = head
    fast = head&.next

    while !fast.nil? and !slow.nil? do
        return true if slow == fast

        slow = slow&.next
        fast = fast&.next&.next
    end

    return false
end


require 'set'

# @param {ListNode} head
# @return {Boolean}
def hasCycle_memory(head)
    cache = Set.new

    while !head.nil? do
        return true if cache.include?(head)

        cache << head
        head = head.next
    end
    
    return false
end